src/XliffTasks/Model/Document.cs

```diff
@@ -62,12 +62,34 @@ public void Load(Stream stream)
         /// </summary>
         public void Save(string path)
         {
+
+            // On Windows:
+            // Readers will prevent the file from being overwritten due to FileShare.Read.
+            // Readers can read in parallel, but when there's contention with a writer
+            // On Unix:
+            // FileShare.Read does nothing, but...
+            // File.Replace is implemented with rename system call that will mean that even though writers can overwrite while
+            // reading is happening, each reader will see file before or after overwrite, not in between, which is the same reader-writer semantic we want.
             EnsureContent();
+            var tempPath = Path.Combine(Path.GetDirectoryName(path), Path.GetRandomFileName());
 
-            using (var stream = File.Open(path, FileMode.Create, FileAccess.ReadWrite, FileShare.None))
+            using (var stream = File.Open(tempPath, FileMode.Create, FileAccess.ReadWrite, FileShare.None))
             {
                 Save(stream);
             }
+
+            try
+            {
+                File.Move(tempPath, path);
+            }
+            catch (IOException e) when (e.Message.Contains("Cannot create a file when that file already exists"))
```

Comment:
Exception messages are localized so we can't check it like this.

Reply:
I think this should be 

``` C#
if (File.Exists(tempPath))
{
   File.Replace(...)
}
else
{
    File.Move(...)
}
```

With try/catch (IOException) around the whole thing with retry.

Reply:
> Exception messages are localized

You are right...

> if (File.Exists(tempPath))

I used to avoid read before write when I work on database, since there will be a concurrency issue if something happen in the middle of File.Exists and File.Replace. I am not sure about file system. However, if we add Retry around it, there should not be a problem then. And I will do that

Reply:
I agree about exists checks in general. It's unfortunate that we have to use different API for when the file exists vs. doesn't here. I filed https://github.com/dotnet/corefx/issues/20755 about that and other issues related to this. 

The exists check here, if it were inside try/catch/retry as described above would not be incorrect. Even if there were a race between !exists and move, we'd retry and then replace.

Reply:
fixed